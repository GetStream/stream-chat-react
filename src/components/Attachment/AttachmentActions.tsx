import React from 'react';
import type { Action, Attachment } from 'stream-chat';

import type { ActionHandlerReturnType } from '../Message/hooks/useActionHandler';
import type { DefaultStreamChatGenerics } from '../../types/types';

export type AttachmentActionsProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = Attachment<StreamChatGenerics> & {
  /** A list of actions */
  actions: Action[];
  /** Unique id for action button key. Key is generated by concatenating this id with action value - {`${id}-${action.value}`} */
  id: string;
  /** The text for the form input */
  text: string;
  /** Click event handler */
  actionHandler?: ActionHandlerReturnType;
};

const UnMemoizedAttachmentActions = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: AttachmentActionsProps<StreamChatGenerics>,
) => {
  const { actionHandler, actions, id, text } = props;

  const handleActionClick = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    name?: string,
    value?: string,
  ) => actionHandler?.(name, value, event);

  return (
    <div className='str-chat__message-attachment-actions'>
      <div className='str-chat__message-attachment-actions-form'>
        <span>{text}</span>
        {actions.map((action) => (
          <button
            className={`str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--${action.style}`}
            data-testid={`${action.name}`}
            data-value={action.value}
            key={`${id}-${action.value}`}
            onClick={(event) => handleActionClick(event, action.name, action.value)}
          >
            {action.text}
          </button>
        ))}
      </div>
    </div>
  );
};

/**
 * A component for rendering the actions you can take on an attachment.
 */
export const AttachmentActions = React.memo(
  UnMemoizedAttachmentActions,
) as typeof UnMemoizedAttachmentActions;
