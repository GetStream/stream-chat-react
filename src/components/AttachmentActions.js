import React from 'react';
import PropTypes from 'prop-types';

/**
 * AttachmentActions - The actions you can take on an attachment
 *
 * @example ./docs/AttachmentActions.md
 * @extends PureComponent
 */
export class AttachmentActions extends React.PureComponent {
  static propTypes = {
    // /** The id of the form input */
    // id: PropTypes.string.isRequired,
    /** The text for the form input */
    text: PropTypes.string,
    /** A list of actions */
    actions: PropTypes.array.isRequired,
    /** The handler to execute after selecting an action */
    actionHandler: PropTypes.func.isRequired,
  };

  render() {
    const { text, id, actions, actionHandler } = this.props;

    return (
      <div className="str-chat__message-attachment-actions">
        <form className="str-chat__message-attachment-actions-form">
          <span key={0}>{text}</span>
          {actions.map((action) => (
            <button
              className={`str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--${action.style}`}
              key={`${id}-${action.value}`}
              data-value={action.value}
              onClick={actionHandler.bind(this, action.name, action.value)}
            >
              {action.text}
            </button>
          ))}
        </form>
      </div>
    );
  }
}
