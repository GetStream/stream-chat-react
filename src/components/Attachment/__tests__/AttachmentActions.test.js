import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, waitFor } from '@testing-library/react';

import renderer from 'react-test-renderer';
import { nanoid } from 'nanoid';
import { AttachmentActions } from '../AttachmentActions';

const getComponent = (props) => <AttachmentActions {...props} />;
const actions = [
  {
    name: 'action 1 name',
    text: 'action 1 text',
    value: 'action 1',
  },
  {
    name: 'action 2 name',
    text: 'action 2 text',
    value: 'action 2',
  },
];

describe('AttachmentActions', () => {
  it('should render AttachmentActions component', () => {
    const tree = renderer
      .create(
        getComponent({
          actionHandler: jest.fn(),
          actions,
          id: nanoid(),
        }),
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should call actionHandler on click', async () => {
    const actionHandler = jest.fn();
    const { getByTestId } = render(
      getComponent({
        actionHandler,
        actions,
        id: nanoid(),
      }),
    );

    await waitFor(() => {
      expect(getByTestId(actions[0].name)).toBeInTheDocument();
    });

    fireEvent.click(getByTestId(actions[0].name));
    fireEvent.click(getByTestId(actions[1].name));

    await waitFor(() => {
      // eslint-disable-next-line jest/prefer-called-with
      expect(actionHandler).toHaveBeenCalledTimes(2);
    });
  });
});
