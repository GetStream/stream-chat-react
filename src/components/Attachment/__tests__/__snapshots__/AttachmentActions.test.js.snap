// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentActions should render AttachmentActions component 1`] = `
<div>
  <div
    class="str-chat__message-attachment-actions"
  >
    <div
      class="str-chat__message-attachment-actions-form"
    >
      <span />
      <button
        class="str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--undefined"
        data-testid="action 1 name"
        data-value="action 1"
      >
        action 1 text
      </button>
      <button
        class="str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--undefined"
        data-testid="action 2 name"
        data-value="action 2"
      >
        action 2 text
      </button>
    </div>
  </div>
</div>
`;
